src/descriptor/satisfied_contraints.rs

```diff
@@ -0,0 +1,1482 @@
+    PublicKey {
+        key: &'desc Pk,
+        sig: secp256k1::Signature,
+    },
+    ///Public keyhash, corresponding pubkey and signature
+    PublicKeyHasH {
+        keyhash: &'desc Pkh,
+        key: bitcoin::PublicKey,
+        sig: secp256k1::Signature,
+    },
+    ///Hashlock and preimage for SHA256
+    HashLock {
+        hash: HashLockType<'desc>,
+        preimage: &'stack [u8],
+    },
+    ///Relative Timelock for CSV.
+    RelativeTimeLock {
+        time: &'desc u32,
+    },
+    ///Absolute Timelock for CLTV.
+    AbsoluteTimeLock {
+        time: &'desc u32,
+    },
+}
+
+///This is used by the interpreter to know which evaluation state a AstemElem is.
+///This is required because whenever a same node(for eg. OrB) appears on the stack, we don't
+///know if the left child has been evaluated or not. And based on the result on
+///the top of the stack, we need to decide whether to execute right child or not.
+///This is also useful for wrappers and thresholds which push a value on the stack
+///depending on evaluation of the children.
+enum NodeEvaluationState{
+    ///No children have been evaluted
+    NoChildrenCompleted,
+    ///Only the left child has been evaluated
+    LeftChildCompleted,
+    ///Only partial evaluation completed(used for Thresh nodes)
+    PartialThreshCompleted,
+    ///Multi-sig verification which represents currently k out n are satisfied.
+    ///This is required because the children of multi are not AstElem and
+    ///cannot be interpreted directly. Thus, we need to keep track of them
+    ///separately
+    PartialMultiCompleted(usize,usize),
+    ///All chilren have been evaluated
+    AllChildrenCompleted
+}
+
+/// An iterator over all the constraints satisfied by a given
+/// descriptor/scriptSig/witness stack tuple
+pub struct SatisfiedConstraints<'desc, 'stack, Pk: 'desc, Pkh: 'desc> {
+    secp: &'desc secp256k1::Secp256k1<VerifyOnly>,
+    sighash: &'desc secp256k1::Message,
+    public_key: Option<&'desc Pk>,
+    state : Vec <(&'desc AstElem<Pk, Pkh>, NodeEvaluationState)>,
+    stack : Vec<StackElement<'stack>>,
+    altstack: Vec<StackElement<'stack>>,
+    age: &'desc u32,
+    height: &'desc u32,
+}
+
+impl<'desc, 'stack, Pk, Pkh> SatisfiedConstraints<'desc, 'stack, Pk, Pkh>{
+
+    pub fn from_desriptor(
+        secp: &'desc secp256k1::Secp256k1<VerifyOnly>,
+        sighash: &'desc secp256k1::Message,
+        des: &'desc Descriptor<Pk, Pkh>,
+        stack: Vec<StackElement<'stack>>,
+        age: &'desc u32,
+        height: &'desc u32,
+    ) -> SatisfiedConstraints<'desc, 'stack, Pk, Pkh> where
+    Pk: ToPublicKey, Pkh: ToPublicKeyHash
+    {
+        match des {
+            &Descriptor::Pk(ref pk) |
+            &Descriptor::Pkh(ref pk) |
+            &Descriptor::ShWpkh(ref pk) |
+            &Descriptor::Wpkh(ref pk) => SatisfiedConstraints{
+                secp,
+                sighash,
+                public_key: Some(&pk),
```

Comment:
Why not just put a `Miniscript(AstNode::Pk(pk))` onto the stack rather than special-casing this?

Reply:
I cannot move pk into `AstElem::Pk(pk)` without cloning it explicitly. Our `AstElems::Pk` hold pk not references to them.

Reply:
Ah. Ok, then I think you should just clone the `pk`.

Reply:
@apoelstra , I don't think I can avoid the special case here. The function expects the state vector object to have references to AstElems with lifetimes outside of the function scope. I cannnot use anything created in the function along with it

Reply:
Yep, I think you're right.

Reply:
Another way to avoid this is to change the descriptors `Pk`, `Pkh`, `Wpkh` and `ShWpkh` from Pk to accept Miniscript. That is Pk(Pk) --> Pk(Miniscript<Pk, Pkh>).

Reply:
No, I think the existing solution is fine